### ISSUE TITLE
pcloud crashes with SIGSEGV on recent nixpkgs unstable

### ISSUE THREAD

**Issue by r-vdp** [NixOS/nixpkgs#226339]
### Describe the bug
Running pcloud from nixos-unstable crashes with a segmentation fault:

```
$ NIXPKGS_ALLOW_UNFREE=1 nix run --impure 'github:nixos/nixpkgs/nixos-unstable#pcloud'
fish: Job 1, 'NIXPKGS_ALLOW_UNFREE=1 nix run …' terminated by signal SIGSEGV (Address boundary error)
```

The same build used to work before, so I guess there is an ABI incompatibility in one of the runtime dependencies.

**Comment by naturallaw777** [comment]
Having the same issue with NixOS unstable on my machine.

**Comment by juanibiapina** [comment]
Does anyone know a version of nixpkgs I can pin where pcloud is still working?

**Comment by juanibiapina** [comment]
@Patryk27 hey, sorry to tag you, but maybe you can help ?

**Comment by Fuzen-py** [comment]
> Does anyone know a version of nixpkgs I can pin where pcloud is still working?
You can try following the revisions in this [flake](https://github.com/chenlijun99/dotfiles/commit/30b6f8c2f32b5321544f6f84c078d33c227574da).

**Comment by R-VdP** [comment]
@juanibiapina this is what works for me currently: https://git.sr.ht/~r-vdp/nixos-config/tree/main/item/flake.nix#L131

**Comment by Fuzen-py** [comment]
> @juanibiapina this is what works for me currently: https://git.sr.ht/~r-vdp/nixos-config/tree/main/item/flake.nix#L131

Unable to figure out why your hash is different, it seems to be the same archive in [nixpkgs](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/pcloud/default.nix#L117)

**Comment by Patryk27** [comment]
> @Patryk27 hey, sorry to tag you, but maybe you can help ?

Thanks for tagging - I'll take a look later today or tomorrow!

**Comment by Patryk27** [comment]
Interestingly, it crashes somewhere in icu-60:

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff707aa03 in icu_60::UnicodeString::copyFrom(icu_60::UnicodeString const&, signed char) () from /nix/store/hspijmy7kfrs0f12h7b56f07lnx3522v-pcloud-1.12.0/app/libnode.so

(gdb) bt
#0  0x00007ffff707aa03 in icu_60::UnicodeString::copyFrom(icu_60::UnicodeString const&, signed char) () from /nix/store/hspijmy7kfrs0f12h7b56f07lnx3522v-pcloud-1.12.0/app/libnode.so
#1  0x00007ffff6ed494e in icu_60::BasicTimeZone::BasicTimeZone(icu_60::UnicodeString const&) () from /nix/store/hspijmy7kfrs0f12h7b56f07lnx3522v-pcloud-1.12.0/app/libnode.so
#2  0x00007ffff6f7bf55 in icu_60::SimpleTimeZone::SimpleTimeZone(int, icu_60::UnicodeString const&) () from /nix/store/hspijmy7kfrs0f12h7b56f07lnx3522v-pcloud-1.12.0/app/libnode.so
#3  0x00007ffff6f78feb in icu_60::TimeZone::detectHostTimeZone() () from /nix/store/hspijmy7kfrs0f12h7b56f07lnx3522v-pcloud-1.12.0/app/libnode.so
#4  0x00007ffff6f7909f in icu_60::TimeZone::createDefault() () from /nix/store/hspijmy7kfrs0f12h7b56f07lnx3522v-pcloud-1.12.0/app/libnode.so
#5  0x00000000030c8ac5 in ?? ()
#6  0x000000000387f733 in ?? ()
#7  0x000000000247bfa6 in ?? ()
#8  0x0000000003880394 in ?? ()
#9  0x0000000004aeae46 in main ()
```

... but faking a timezone:

```
(gdb) set env TZ=xd
(gdb) r
```

... causes it to both print a weird warning message:

```
[2774010:0517/180552.216458:FATAL:memory_linux.cc(36)] Out of memory.
```

... and segfault somewhere else - during v8's initialization:

... 
Thread 1 "pcloud" received signal SIGSEGV, Segmentation fault.
0x00007ffff783b072 in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
(gdb) bt
#0  0x00007ffff783b072 in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#1  0x00007ffff783efdf in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#2  0x00007ffff783f03e in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#3  0x00007ffff70d84e7 in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#4  0x00007ffff762842b in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#5  0x00007ffff6e633ef in v8::base::CallOnceImpl(long*, std::__1::function<void ()>) () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#6  0x00007ffff76283b0 in ?? () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#7  0x00007ffff7802419 in v8::V8::Initialize() () from /nix/store/rhlr6x0199x6xj02cywqidp4f8vcq71y-pcloud-1.12.0/app/libnode.so
#8  0x000000000393411b in ?? ()
#9  0x000000000393114b in ?? ()
#10 0x0000000004b418cb in atom::JavascriptEnvironment::Initialize() ()
#11 0x0000000004b41731 in atom::JavascriptEnvironment::JavascriptEnvironment() ()
#12 0x0000000004b3800e in atom::AtomBrowserMainParts::PostEarlyInitialization() ()
#13 0x00000000035a28d7 in ?? ()
#14 0x00000000035d0a9a in ?? ()
#15 0x00000000035a1c7f in ?? ()
#16 0x000000000387fb78 in ?? ()
#17 0x000000000247c414 in ?? ()
#18 0x0000000003880394 in ?? ()
#19 0x0000000004aeae46 in main ()
```

This weird-warning-message and crashes don't happen when I check out older nixpkgs (e.g. `git checkout a575c243c23e2851b78c00e9fa245232926ec32f`) and apply newest pcloud onto it - in that case everything just works which suggests an ABI mismatch on the libnode.so <-> nixpkgs' icu boundary (maybe we're linking stuff that accidentally produces `UnicodeString` using e.g. icu-70 and passes instances of that class into libnode.so, not sure yet).

I'm trying to bisect what's changed but it takes a while 😅

**Comment by naturallaw777** [comment]
Thanks so much @Patryk27 for looking into this!

**Comment by Patryk27** [comment]
Okie, it looks like https://github.com/NixOS/nixpkgs/pull/209870 is somehow at fault here:

```
$ git checkout fdd49f1bcd8a7f0b5e29f550d698b2abe5c540cd
$ NIXPKGS_ALLOW_UNFREE=1 nix run --impure .#pcloud
# seems to work
```

```
$ git checkout 5f57c2e0f97a83bf5691ac3a29da6ef9b44535c4
$ NIXPKGS_ALLOW_UNFREE=1 nix run --impure .#pcloud
# Segmentation fault
```

I'm not yet sure what's actually causing it, though 🤔

**Comment by Patryk27** [comment]
More funky stuff: it looks like the first divergence happens on a `readlink()` somewhere in `uprv_tzname_60()` -- if I put a breakpoint there, the correct version has:

```
(gdb) print (char*)$rdi
$1 = 0x7ffff6abb9b6 "/etc/localtime"
(gdb) bt
#0  0x00007ffff4afff49 in readlink () from /nix/store/59nni10pyifs8mdmdxibqqihk87mk0w6-glibc-2.35-224/lib/libc.so.6
#1  0x00007ffff742f564 in uprv_tzname_60 () from /nix/store/qnx9ryj7qdmg5mnb7gb597dqimwh6jpv-pcloud-1.11.0/app/libnode.so
#2  0x00007ffff7378eeb in icu_60::TimeZone::detectHostTimeZone() () from /nix/store/qnx9ryj7qdmg5mnb7gb597dqimwh6jpv-pcloud-1.11.0/app/libnode.so
#3  0x00007ffff737909f in icu_60::TimeZone::createDefault() () from /nix/store/qnx9ryj7qdmg5mnb7gb597dqimwh6jpv-pcloud-1.11.0/app/libnode.so
#4  0x00000000030c8ac5 in ?? ()
#5  0x000000000387f733 in ?? ()
#6  0x000000000247bfa6 in ?? ()
#7  0x0000000003880394 in ?? ()
#8  0x0000000004aeae46 in main ()
```

... while the invalid version does:

```
(gdb) print (char*)$rdi
$1 = 0x7ffff66bb9b6 'X' <repeats 200 times>...
(gdb) bt
#0  0x00007ffff46b9f69 in readlink () from /nix/store/yaz7pyf0ah88g2v505l38n0f3wg2vzdj-glibc-2.37-8/lib/libc.so.6
#1  0x00007ffff702f564 in uprv_tzname_60 () from /nix/store/0x8qmm4klmi2rzfia6ykin5wcn9ai27j-pcloud-1.12.0/app/libnode.so
#2  0x00007ffff6f78eeb in icu_60::TimeZone::detectHostTimeZone() () from /nix/store/0x8qmm4klmi2rzfia6ykin5wcn9ai27j-pcloud-1.12.0/app/libnode.so
#3  0x00007ffff6f7909f in icu_60::TimeZone::createDefault() () from /nix/store/0x8qmm4klmi2rzfia6ykin5wcn9ai27j-pcloud-1.12.0/app/libnode.so
#4  0x00000000030c8ac5 in ?? ()
#5  0x000000000387f733 in ?? ()
#6  0x000000000247bfa6 in ?? ()
#7  0x0000000003880394 in ?? ()
#8  0x0000000004aeae46 in main ()
```

**Comment by Patryk27** [comment]
Okie, I think I've found the culprit - patchelf generates an invalid `libnode.so` file, which is probably caused by https://github.com/NixOS/patchelf/issues/482.

Downgrading patchelf doesn't help here though, and that kinda suggests that there exist some another issue in patchelf that has just gotten exacerbated by https://github.com/NixOS/nixpkgs/pull/209870.

No fix just yet, although I'm open to ideas 😅

**Comment by StefanSchroeder** [comment]
Could you offer a workaround while this patchelf-error is not fixed? For non-flake users?

**Comment by R-VdP** [comment]
> Could you offer a workaround while this patchelf-error is not fixed? For non-flake users?

@StefanSchroeder you can do essentially the same as I did, but instead of getting the 22.11 nixpkgs through a flake input, you get it with `fetchTarball`.

**Comment by StefanSchroeder** [comment]
> > Could you offer a workaround while this patchelf-error is not fixed? For non-flake users?
> 
> @StefanSchroeder you can do essentially the same as I did, but instead of getting the 22.11 nixpkgs through a flake input, you get it with `fetchTarball`.

Ummmh, my Nix-foo is not that well developed. Would you guide me through it? I installed 23.05 yesterday. I can examine and patch pcloud/default.nix. But as far as I understand, the error is outside of pcloud. So, where exactly would I put the `fetchTarball`?

**Comment by R-VdP** [comment]
> > > Could you offer a workaround while this patchelf-error is not fixed? For non-flake users?
> > 
> > 
> > @StefanSchroeder you can do essentially the same as I did, but instead of getting the 22.11 nixpkgs through a flake input, you get it with `fetchTarball`.
> 
> Ummmh, my Nix-foo is not that well developed. Would you guide me through it? I installed 23.05 yesterday. I can examine and patch pcloud/default.nix. But as far as I understand, the error is outside of pcloud. So, where exactly would I put the `fetchTarball`?

@StefanSchroeder a basic NixOS module could look something like this: https://gist.github.com/R-VdP/999dd803c96aee0cb6a176eadef0978e

I included the part that updates pcloud to the latest 1.13.0 release.

**Comment by StefanSchroeder** [comment]
> @StefanSchroeder a basic NixOS module could look something like this: https://gist.github.com/R-VdP/999dd803c96aee0cb6a176eadef0978e
> 
> I included the part that updates pcloud to the latest 1.13.0 release.

When I import that file `pcloud.nix` from my configuration.nix, I get this error:

```
error: anonymous function at /etc/nixos/pcloud.nix:1:1 called with unexpected argument 'config'

       at /nix/var/nix/profiles/per-user/root/channels/nixos/lib/modules.nix:519:8:

          518|       # works.
          519|     in f (args // extraArgs);
             |        ^
          520|
(use '--show-trace' to show detailed location information)
building Nix...
error: anonymous function at /etc/nixos/pcloud.nix:1:1 called with unexpected argument 'config'

       at /nix/var/nix/profiles/per-user/root/channels/nixos/lib/modules.nix:519:8:

          518|       # works.
          519|     in f (args // extraArgs);
             |        ^
          520|
(use '--show-trace' to show detailed location information)
building the system configuration...
error: anonymous function at /etc/nixos/pcloud.nix:1:1 called with unexpected argument 'config'

       at /nix/var/nix/profiles/per-user/root/channels/nixos/lib/modules.nix:519:8:

          518|       # works.
          519|     in f (args // extraArgs);
             |        ^
          520|
(use '--show-trace' to show detailed location information)

```

**Comment by Patryk27** [comment]
If you're doing:

```
imports = [
    ./that-file.nix
];
```

... then you'll have to change `config.environment.systemPackages` into just `environment.systemPackages` and, possibly, change `{ pkgs, lib }:` into `{ pkgs, lib, ... }:` - then the expression should work 🙂

**Comment by StefanSchroeder** [comment]
That worked!

**Comment by R-VdP** [comment]
Ah yeah, that should have been `{ lib, pkgs, ... }` indeed.

**Comment by naturallaw777** [comment]
Any news on when this will be merged? Thanks everyone!

**Comment by Patryk27** [comment]
No idea, unfortunately - it looks like patchelf has an unresolved bug so it's not just a matter of waiting for when something gets merged somewhere, but rather waiting for someone to fix that bug; I don't have enough time on my hands to taking a stab at it, though 👀

**Comment by naturallaw777** [comment]
Thanks for the response and all the work that you contributed! Right now my skill set is not very high in this area, I wish I could help more. Maybe in the future as I lean more.

**Comment by R-VdP** [comment]
@Patryk27: I had a quick look today, and I noticed that pcloud uses `autoPatchelfHook` which, unless I misread something, doesn't actually use https://github.com/NixOS/patchelf but instead uses https://github.com/eliben/pyelftools.
I wasn't aware that we had multiple implementations of patchelf in NixOS.

So I guess we might need to report the issue in that repo.

It could also be that actually using the `patchelf` tool would not cause the corrupt files, but I didn't have the time to try this yet.

**Comment by Patryk27** [comment]
fwiw, @R-VdP, it looks like that hook uses pyelftools to _scan_ the binary files, but all the patching ultimately ends up happening through patchelf (e.g. https://github.com/NixOS/nixpkgs/blob/724e8ed8d0a108f72e4c42f3bead7601bbeba64d/pkgs/build-support/setup-hooks/auto-patchelf.py#L252)

This also matches my observations in https://github.com/NixOS/patchelf/issues/482#issuecomment-1556159408 where changing the version of patchelf generates files that are broken in different ways, which suggests that something might be going on there :eyes:

**Comment by Patryk27** [comment]
Status: I've gotten back to fixing this bug - currently I've narrowed the issue down to this place:

https://github.com/NixOS/patchelf/blob/7c2f768bf9601268a4e71c2ebe91e2011918a70f/src/patchelf.cc#L1294

... which doesn't implement support for rewriting values other than `STT_SECTION` - in the case of `libnode.so` (which is the reason pcloud doesn't start) the issue is that extending its `RPATH` causes the `.rodata` section to get relocated _without_ updating other references that point into this section (or at least that's my hunch and that's what inspecting with `eu-elflint` suggests).

I think this used not to be a problem before https://github.com/NixOS/nixpkgs/pull/209870 got merged, because that change made `autoPatchelfHook` notice `libnode.so` depends on `libgcc_s.so.1` and so now it tries to include `/nix/store/blahblah-libgcc/lib` in `libnode.so`'s `RPATH`, triggering the unimplemented patching code in `patchelf` (but that change itself is alright, it's `patchelf` which is the culprit here).

tl;dr when building pcloud, patchelf modifies its libraries (shuffling stuff in the memory), but then forgets to update other parts of binary, which continue to refer to the older memory addresses instead of the new ones; this causes sigsegv when trying to start the app

I'll report back once I have more information.

Related:
- https://github.com/NixOS/patchelf/issues/75 (2016, wow!)
- https://github.com/NixOS/patchelf/issues/531
- https://github.com/NixOS/patchelf/issues/482
- https://github.com/NixOS/patchelf/issues/520 (probably, but feels a bit different)

**Comment by philg-dev** [comment]
> > > > Could you offer a workaround while this patchelf-error is not fixed? For non-flake users?
> > > 
> > > 
> > > @StefanSchroeder you can do essentially the same as I did, but instead of getting the 22.11 nixpkgs through a flake input, you get it with `fetchTarball`.
> > 
> > 
> > Ummmh, my Nix-foo is not that well developed. Would you guide me through it? I installed 23.05 yesterday. I can examine and patch pcloud/default.nix. But as far as I understand, the error is outside of pcloud. So, where exactly would I put the `fetchTarball`?
> 
> @StefanSchroeder a basic NixOS module could look something like this: https://gist.github.com/R-VdP/999dd803c96aee0cb6a176eadef0978e
> 
> I included the part that updates pcloud to the latest 1.13.0 release.

Thanks a lot for providing this workaround.

Just in case anybody else still needs to use this workaround...
I've updated the pCloud version used by this workaround to the current `1.14.5` from the pCloud website as follows:

```
      version = "1.14.5";
      code = "XZ0AMJ0ZdrENNeVMNI4Tz3lO1nxr577ryOMV";
      # Archive link's codes: https://www.pcloud.com/release-notes/linux.html
      src = pkgs.fetchzip {
        url = "https://api.pcloud.com/getpubzip?code=${code}&filename=${prev.pname}-${version}.zip";
        hash = "sha256-a577iWPrke3EizG03m0+hjSoPzA4wDai/QMX2Zl7MF0=";
      };
```

**Comment by StefanSchroeder** [comment]
Even though the title of this issue no longer applies, because I am reporting for 24.05-stable here, 
pcloud still continuous to segfault w/ version 1.14.5. Tried with recently released stable version.

**Comment by VanDoge** [comment]
> Even though the title of this issue no longer applies, because I am reporting for 24.05-stable here, pcloud still continuous to segfault w/ version 1.14.5. Tried with recently released stable version.

can confirm issue persists on 24.05

**Comment by StefanSchroeder** [comment]
@Patryk27 you had done some work on this. It seems that https://github.com/NixOS/patchelf/pull/544 might be the solution.
It's open for a while, has seen some review and improvements. Is there any way to push this forward?

**Comment by Patryk27** [comment]
So, the remaining issue is a couple of tests failing on less popular architectures - if you scroll down to the bottom of [the pull request](https://github.com/NixOS/patchelf/pull/544), right after `Some checks were not successful`, you'll see that the tests succeed on x86, but fail for ppc64le, arm64v8 etc - it's possibly related to me misunderstanding some piece of logic:

https://github.com/NixOS/patchelf/pull/544#discussion_r1483322623

I haven't had much time to re-investigate it, but I _do_ have a couple of spare hours this weekend, so I'll try giving it another shot.

That being said, you can still [apply the patch on your system](https://discourse.nixos.org/t/pcloud-gives-segmentation-fault/31330/5?u=patryk27) in the meantime, you don't have to wait until it's merged.

**Comment by gitmeED331** [comment]
NixOX 24.05 
pcloud is still segfault on attempting to run it

**Comment by Patryk27** [comment]
Yes, the problem is that patchelf's maintainers haven't had the capacity to review my changes yet; I'll try asking around again.

**Comment by Patryk27** [comment]
Alright, https://github.com/NixOS/patchelf/pull/544 got merged! The next step is bumping patchelf within nixpkgs and we should be done with this bug 🙂

**Comment by boustanihani** [comment]
Still having the same issue using` v1.14.8` on `NixOS 24.11` =>

`Segmentation fault (Speicherabzug geschrieben)`

**Comment by philg-dev** [comment]
@boustanihani well, what do you expect by using a pcloud version that is from before the fix for this issue got merged?

Both the pcloud version bump from 1.14.8 to 1.14.9 and the patchelf fix got merged on January 25th, so if you're on 1.14.8 you simply don't have the patchelf fix yet either.
